### PR TITLE
Defined classes Customer and Order. Added methods total, add_product,…

### PR DESCRIPTION
**Pull request by snowistaken**
… and remove_product for class Order.

# Assignment Submission: Grocery Store

Congratulations! You're submitting your assignment. Please reflect on the assignment with these questions.

## Reflection

Question | Response
--- | ---
What is accomplished with `raise ArgumentError`? | The program is stopped and the error and its corresponding message is returned to the console. This helps prevent the program continuing operation when something that may cause problems occurs.
Why do you think we made the `.all` & `.find` methods class methods?  Why not instance methods? | .all and .find are class methods because when we call those methods we want to search/return values from all of the instances of that class.
Think about the relation between `Order` and `Customer`. Is this relation one-to-one, one-to-many, or something else? How does that compare to the Solar System project? | The relation is one to many, as one customer can have many orders. This compares similarly to the solar system in that we had one solar system for many planets.
How is the relation between `Order` and `Customer` tracked in the CSV file? How is it tracked in your program? Why might these be different? | The relation between order and customer is tracked by including the customer id and all order properties in the same row as one order. This is different in that in the CSV the customer is more "inside" of each order, but in my program each order is "inside" of the customer.
Did the presence of automated tests change the way you thought about the problem? How? | The presence of automated tests definitely forces me to consider edge cases and how my program will perform under standard circumstances. This helps keep my coding process structured and it assists in finding errors.

### CODE REVIEW

File: lib/customer.rb [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+  def initialize(id = 0, email = '', address = {})
+    @id = id
+    @email = email
+    @address = address
+  end
+
+  def self.all
+    all_customers = []
+    CSV.read('../data/customers.csv').each do |row|
```
Comment (kaidamasaki) [inline lib/customer.rb:16]: The path you have here makes it so that `rake` won't successfully run tests.  (They work when run individually.)

To make `rake` work you can update the path relative to the project root:
```suggestion
    CSV.read('data/customers.csv').each do |row|
```

It's really unfortunate that this works differently from `require_relative`.

File: lib/order.rb [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    if @products.has_key?(name)
+      @products.delete(name)
+    else
+      raise ArgumentError.new "The product #{name} is not present in the order."
+    end
+  end
+
+  def self.all
+    all_orders = []
+    CSV.read('../data/orders.csv').each do |row|
```
Comment (kaidamasaki) [inline lib/order.rb:41]: See above.

```suggestion
    CSV.read('data/orders.csv').each do |row|
```

File: lib/order.rb [hunk 3]
```diff
@@ -0,0 +1,62 @@
+require 'csv'
+
+class Order
+  attr_reader :id
+  attr_accessor :products, :customer, :fulfillment_status
+
+  def initialize(id, products, customer, fulfillment_status = :pending)
+    valid_status = [:pending, :paid, :processing, :shipped, :complete]
+    unless valid_status.include?(fulfillment_status)
+      raise ArgumentError.new "#{fulfillment_status} is not a valid status."
```
Comment (kaidamasaki) [inline lib/order.rb:10]: :tada: Saying what argument was bad!  :tada:

**Comment by kaidamasaki** [comment]
# Grocery Store





## Major Learning Goals/Code Review



| Criteria | yes/no, and optionally any details/lines of code to reference |
| --- | --- |
The `Customer` and `Order` classes have appropriate `attr_` helpers for instance variables | ✔️
The `CSV` library is used in `.all` methods, and not in `.find` methods | ✔️
Appropriately parses the product data from CSV file in `Order.all` | ✔
`Order.all` calls `Customer.find` to set up the composition relation | ✔️
Practices git with at least 3 small commits and meaningful commit messages |

## Functional Requirements

| Functional Requirement | yes/no |
| --- | --- |
All of the wave 1 tests in `Customer` pass | ✔️
All of the wave 1 tests in `Order` pass | ✔️
All of the wave 2 tests in `Customer` pass | ✔️
All of the wave 2 tests in `Order` pass | Didn't write tests for order wave 2.

## Overall Feedback

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Green (Meets/Exceeds Standards) | 4+ in Code Review && 3+ in Functional Requirements | ✔️
| Yellow (Approaches Standards) | 2+ in Code Review && 2+ in Functional Requirements |
| Red (Not at Standard) | 0,1 in Code Review or 0,1 in Functional Reqs, or assignment is breaking/doesn’t run with less than 5 minutes of debugging |





## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Perfect Indentation |
| Elegant/Clever |
| Descriptive/Readable | ✅
| Concise |
| Logical/Organized | ✅
